charts/hedera-mirror-common/templates/zfs/configmap-init.yaml

```
@@ -7,29 +7,155 @@ metadata:
-    LABELS="$(gcloud compute instances describe "${NODE_NAME}" --zone="${ZONE}" --flatten="metadata.kube-labels[]")"
-    CITUS_ROLE="$(echo "${LABELS}" | grep -o 'citus-role[^, }]*' | sed -n 's/.*citus-role=\(.*\).*/\1/p')"
-    DISK_NAME="${NODE_NAME}-zfs"
-    echo "Setting up disk ${DISK_NAME} for ${CITUS_ROLE}"
+    NODE_NAME="$(curl -sS http://metadata.google.internal/computeMetadata/v1/instance/hostname -H 'Metadata-Flavor: Google' | awk -F'.' '{print $1}')"
+    ZONE="$(curl -sS http://metadata.google.internal/computeMetadata/v1/instance/zone -H 'Metadata-Flavor: Google' | awk -F  "/" '{print $NF}')"
+    
+    # Configure kubectl
+    APISERVER="https://kubernetes.default.svc"
+    SERVICEACCOUNT="/var/run/secrets/kubernetes.io/serviceaccount"
+    TOKEN="$(cat ${SERVICEACCOUNT}/token)"
+    CACERT="${SERVICEACCOUNT}/ca.crt"
+    
+    kubectl config set-cluster ctc --server="${APISERVER}" --certificate-authority="${CACERT}"
+    kubectl config set-context ctc --cluster=ctc
+    kubectl config set-credentials user --token="${TOKEN}"
+    kubectl config set-context ctc --user=user
+    kubectl config use-context ctc
+    
+    NODE_LABELS="$(kubectl get node ${NODE_NAME} -o json | jq '.metadata.labels')"
+    CITUS_ROLE="$(echo "${NODE_LABELS}" | jq --arg ROLE_LABEL "${ROLE_LABEL}" -r '.[$ROLE_LABEL]')"
+    NODE_ID="$(echo "${NODE_LABELS}" | jq --arg NODE_ID_LABEL "${NODE_ID_LABEL}" -r '.[$NODE_ID_LABEL]')"
 
+    ATTEMPT_COUNT=1
+    while [[ "${CITUS_ROLE}" == "null" ]]
+    do
+      if [[ ${ATTEMPT_COUNT} -ge 13 ]]
+      then
+        echo "Timed out waiting for labels please set ${ROLE_LABEL}"
+        exit 1
+      fi
+      sleep 10
+      ATTEMPT_COUNT=$((ATTEMPT_COUNT+1))
+      echo "Retrying label GET attempt: ${ATTEMPT_COUNT}"
+      NODE_LABELS="$(kubectl get node ${NODE_NAME} -o json | jq '.metadata.labels')"
+      CITUS_ROLE="$(echo ${NODE_LABELS} | jq --arg ROLE_LABEL "${ROLE_LABEL}" -r '.metadata.labels[$ROLE_LABEL]')"
+    done
+    # Determine correct config for zone
+    CONFIG_MAP_NAME="zfs-node-status"
+    COMMAND_STATUS=1
+    until [[ ${COMMAND_STATUS} -eq 0 ]]; do
+      CONFIG_MAP="$(kubectl get configmap -o json -n ${K8S_NAMESPACE} ${CONFIG_MAP_NAME})"
+      MODIFIED_CONFIG_MAP="$(echo "${CONFIG_MAP}" |jq '{metadata: {resourceVersion: (.metadata.resourceVersion)}, data: {"zfs-node-status.json": (.data["zfs-node-status.json"]|fromjson)}}')"
+      ZONE_CONFIG="$(echo "${MODIFIED_CONFIG_MAP}" |jq  --arg ZONE "${ZONE}" --arg CITUS_ROLE "${CITUS_ROLE}" '.data["zfs-node-status.json"][$CITUS_ROLE][$ZONE] // []')"
+      NODE_LIST="$(kubectl get nodes -o json | jq \
+                   --arg ZONE "${ZONE}" \
+                   --arg CITUS_ROLE "${CITUS_ROLE}" \
+                   --arg ROLE_LABEL "${ROLE_LABEL}" \
+                   '.items | map(select(.metadata.labels["topology.kubernetes.io/zone"] == $ZONE and
+                                        .metadata.labels[$ROLE_LABEL] == $CITUS_ROLE))')"
```

We should use kubectl selectors to filter on the server where we can. Also seems like we only need the metadata so can return a subset.

```suggestion
      NODE_LIST="$(kubectl get nodes -lcsi-type=zfs,topology.kubernetes.io/zone=${ZONE},${ROLE_LABEL}=${CITUS_ROLE} -o jsonsonpath='{.items[*].metadata}')"
```

---

It would result in less data being returned in cases where we have nodes with a citus-role but not csi-type=zfs which isn't really planned to be the case. I can add the csi-type filter as well. Was mainly just left off to not have more jq logic than necessary but since you have provided other input on moving these filters out of the jq script and into kubectl args I will add this one as well.

+1 on jsonpath. Haven't worked with that before and have always just used jq.